### ISSUE TITLE
We were using InferOpt wrong

### ISSUE THREAD

**Issue by gdalle** [YasBroek/PRE_Internship#10]
I just figured out what went wrong with your use of InferOpt[^1]: the library assumes that you provide a linear *maximizer* $\arg\max_y \theta^\top y$, but independent shortest paths is a *minimizer* $\arg\min_y \theta^\top y = \arg\max_y (-\theta)^\top y$. This means we need to work with $-\theta$ instead of $\theta$ (as shown in the [tutorial](https://juliadecisionfocusedlearning.github.io/InferOpt.jl/stable/tutorial/)). Fixing that does yield a decreasing loss, and it will also generalize to features (as long as you make sure that your linear combination outputs something of the right sign):

[^1]: disregard the manual code I had given you earlier, it had another bug that made it invalid

```julia
using MultiAgentPathFinding  # version from main branch
using MultiAgentPathFinding: replace_weights, vectorize_weights
using SimpleWeightedGraphs
using StableRNGs
using UnicodePlots
using InferOpt
using Zygote

using SparseArrays
using LinearAlgebra
using Statistics

function count_edge_visits(solution::Solution, mapf::MAPF)
    visits = similar(mapf.graph.weights, Int)
    visits.nzval .= 0
    for path in solution.paths
        for t in 1:(length(path) - 1)
            visits[path[t], path[t + 1]] += 1
            if path[t] != path[t + 1]
                visits[path[t + 1], path[t]] += 1
            end
        end
    end
    return vectorize_weights(SimpleWeightedGraph(visits))
end

function maximizer(minus_theta; mapf)
    theta = -minus_theta  # independent_dijkstra is a *minimizer*
    theta_projected = max.(theta, 1e-3)  # ensure positive weights
    new_graph = replace_weights(mapf.graph, theta_projected)
    new_mapf = MAPF(
        new_graph,
        mapf.departures,
        mapf.arrivals;
        vertex_conflicts=mapf.vertex_conflicts,
        edge_conflicts=mapf.edge_conflicts,
    )
    solution_indep = independent_dijkstra(new_mapf)
    y = count_edge_visits(solution_indep, mapf)
    return y
end

grid = Bool[
    1 0 1
    0 0 1
    0 0 0
    0 0 1
]
x1 = (3, 3)
x2 = (2, 1)
y1 = (4, 2)
y2 = (1, 2)
departure_coords = [x1, y1]
arrival_coords = [x2, y2]
mapf = MAPF(grid, departure_coords, arrival_coords; allow_diagonal_moves=false)
_, coord_to_vertex, vertex_to_coord = parse_benchmark_map(grid; allow_diagonal_moves=false)

optimal_solution = cooperative_astar(mapf)  # by accident
y_opt = count_edge_visits(optimal_solution, mapf)

nb_samples = 10
ε = 1e-1
learning_rate = 1e-2

layer = PerturbedAdditive(maximizer; ε, nb_samples, seed=0, rng=StableRNG(0), threaded=true)
loss = FenchelYoungLoss(layer)

minus_theta = -vectorize_weights(mapf.graph)
losses = Float64[]

for epoch in 1:20
    @info "epoch $epoch"
    yield()

    l, grads = Zygote.withgradient(mt -> loss(mt, y_opt; mapf), minus_theta)
    grad = grads[1]

    minus_theta -= learning_rate * grad
    minus_theta = min.(minus_theta, -1e-3)
    push!(losses, l)
end
```

```julia
julia> lineplot(eachindex(losses), losses)
       ┌────────────────────────────────────────┐ 
   1.1 │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
       │⠀⠀⢄⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
       │⠀⠀⠀⠑⠢⣀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
       │⠀⠀⠀⠀⠀⠀⠉⠢⢄⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
       │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⠒⠤⣀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
       │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠉⠒⠤⣀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
       │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠑⠢⢄⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
       │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⠑⠢⢄⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
       │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⠑⠒⠤⣀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
       │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠉⠑⠢⢄⡀⠀⠀⠀⠀⠀⠀│ 
       │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⠉⠒⠤⣀⡀⠀│ 
       │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⠑│ 
       │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
       │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
   0.7 │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
       └────────────────────────────────────────┘ 
       ⠀0⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀20⠀
```
